storage/src/nodestore/alloc.rs

```diff
@@ -20,262 +20,29 @@
-    #[inline]
-    #[must_use]
-    pub const fn new(addr: u64) -> Option<Self> {
-        match NonZeroU64::new(addr) {
-            Some(addr) => Some(LinearAddress(addr)),
-            None => None,
-        }
-    }
-
-    /// Get the underlying address as u64.
-    #[inline]
-    #[must_use]
-    pub const fn get(self) -> u64 {
-        self.0.get()
-    }
-
-    /// Check if the address is 8-byte aligned.
-    #[inline]
-    #[must_use]
-    pub const fn is_aligned(self) -> bool {
-        self.0.get() % (Self::MIN_AREA_SIZE) == 0
-    }
-
-    /// The maximum area size available for allocation.
-    pub const MAX_AREA_SIZE: u64 = *AREA_SIZES.last().unwrap();
-
-    /// The minimum area size available for allocation.
-    pub const MIN_AREA_SIZE: u64 = *AREA_SIZES.first().unwrap();
-
-    /// Returns the number of different area sizes available.
-    #[inline]
-    #[must_use]
-    pub const fn num_area_sizes() -> usize {
-        const { AREA_SIZES.len() }
-    }
-
-    /// Returns the inner `NonZeroU64`
-    #[inline]
-    #[must_use]
-    pub const fn into_nonzero(self) -> NonZeroU64 {
-        self.0
-    }
-
-    /// Advances a `LinearAddress` by `n` bytes.
-    ///
-    /// Returns `None` if the result overflows a u64
-    /// Some(LinearAddress) otherwise
-    ///
-    #[inline]
-    #[must_use]
-    pub const fn advance(self, n: u64) -> Option<Self> {
-        match self.0.checked_add(n) {
-            // overflowed
-            None => None,
-
-            // It is impossible to add a non-zero positive number to a u64 and get 0 without
-            // overflowing, so we don't check for that here, and panic instead.
-            Some(sum) => Some(LinearAddress(sum)),
-        }
-    }
-
-    /// Returns the number of bytes between `other` and `self` if `other` is less than or equal to `self`.
-    /// Otherwise, returns `None`.
-    #[inline]
-    #[must_use]
-    pub const fn distance_from(self, other: Self) -> Option<u64> {
-        self.0.get().checked_sub(other.0.get())
-    }
-}
-
-impl std::ops::Deref for LinearAddress {
-    type Target = NonZeroU64;
-    fn deref(&self) -> &Self::Target {
-        &self.0
-    }
-}
-
-impl fmt::Display for LinearAddress {
-    fn fmt(&self, f: &mut fmt::Formatter<'_>) -> fmt::Result {
-        std::fmt::Display::fmt(&self.get(), f)
-    }
-}
-
-impl fmt::LowerHex for LinearAddress {
-    fn fmt(&self, f: &mut fmt::Formatter<'_>) -> fmt::Result {
-        std::fmt::LowerHex::fmt(&self.get(), f)
-    }
-}
-impl From<LinearAddress> for u64 {
-    fn from(addr: LinearAddress) -> Self {
-        addr.get()
-    }
-}
-
-impl From<NonZeroU64> for LinearAddress {
-    fn from(addr: NonZeroU64) -> Self {
-        LinearAddress(addr)
-    }
-}
-
 /// `FreeLists` is an array of `Option<LinearAddress>` for each area size.
-pub type FreeLists = [Option<LinearAddress>; AREA_SIZES.len()];
+pub type FreeLists = [Option<LinearAddress>; AreaIndex::NUM_AREA_SIZES];
```

Comment:
Looks like `FreeLists` is a candidate for our next newtype! It can have a implementation similar to `Children<T>` and also implement `Index<AreaSize>`.